queryplan_visualization/cmd/querygraph/main.go

```diff
@@ -0,0 +1,232 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// querygraph demonstrates visualizing BigQuery query plans by converting them
+// into graphviz "dot" file notation and using graphviz tooling to emit images.
+package main
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"html"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+
+	"cloud.google.com/go/bigquery"
+)
+
+type jobRef struct {
+	Project  string
+	Location string
+	JobID    string
+}
+
+func (ref *jobRef) FullJobID() string {
+	if ref.Location == "" {
+		return fmt.Sprintf("%s:%s", ref.Project, ref.JobID)
+	}
+	return fmt.Sprintf("%s:%s.%s", ref.Project, ref.Location, ref.JobID)
+}
+
+func parseJobStr(in string) (*jobRef, error) {
+	jr := &jobRef{}
+	i := strings.LastIndex(in, ":")
+	if i <= 0 || len(in) < i+1 {
+		return nil, errors.New("malformed job string")
+	}
+	jr.Project = in[0:i]
+	in = in[i+1:]
+	i = strings.LastIndex(in, ".")
+	if i <= 0 || len(in) < i+1 {
+		// no location split
+		jr.JobID = in
+		return jr, nil
+	}
+	jr.Location = in[0:i]
+	jr.JobID = in[i+1:]
+	return jr, nil
+}
+
+// fetchQueryDetails is responsible for getting the job metadata from BigQuery.
+func fetchQueryDetails(ctx context.Context, ref *jobRef) ([]*bigquery.ExplainQueryStage, error) {
+
+	// construct a BigQuery client using the project referenced in the job reference.
+	client, err := bigquery.NewClient(ctx, ref.Project)
+	if err != nil {
+		return nil, fmt.Errorf("bigquery.NewClient: %v", err)
+	}
+
+	// Fetch job metadata.
+	job, err := client.JobFromIDLocation(ctx, ref.JobID, ref.Location)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get job metadata: %w", err)
+	}
+
+	status, err := job.Status(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get status: %w", err)
+	}
+
+	queryStats, ok := status.Statistics.Details.(*bigquery.QueryStatistics)
+	if !ok {
+		return nil, fmt.Errorf("couldn't retrieve query statistics: %w", err)
+	}
+
+	if len(queryStats.QueryPlan) == 0 {
+		return nil, fmt.Errorf("query plan was empty")
+	}
+	return queryStats.QueryPlan, nil
+
+}
+
+// nameToColor is used for mapping a dominant query stage description to a hex color string.
+func nameToColor(name string) string {
+	parts := strings.Split(name, ":")
+	//remove spaces and plus symbols (indicate complex stages)
+	opType := strings.Trim(parts[1], " +")
+	switch opType {
+	case "Input":
+		return "#f50057"
+	case "Repartition":
+		return "#4caf50"
+	case "Join":
+		return "#ff6f00"
+	default:
+		return "#2196f3"
+	}
+}
+
+// stagesToDot converts the queryDetails datastructure into a dot representation.
+func stagesToDot(stages []*bigquery.ExplainQueryStage, title string) []byte {
+
+	shufOut := make(map[int64]int64)
+	flushOut := make(map[int64]int64)
+	var buf bytes.Buffer
+
+	buf.WriteString("digraph {\n")
```

Comment:
I feel like there has to be a more readable way to build this string. Would the `template` package work for this use case? https://pkg.go.dev/text/template

Reply:
It could certainly be refactored to a more general builder, which I'd do if this got pulled into package for rendering dot files.